README.md
# rst-archive

Recovering data from the Internet Archive snapshot of Reducing Stereotype Threat

## Why?

For years Catherine Good (Baruch College) and Steve Stroessner (Barnard College) maintained `www.reducingstereotypethreat.org`. It consisted of an annotated bibliography of research on stereotype threat, along with discussions of various aspects of stereotype threat (e.g. vulnerable groups, situations, mechanisms, ...). This made it an indispensable resource for learning about stereotype threat and then teaching it to others. When training teaching assistants, I would send them to the site to explore specific questions. This research would prime them for discussion at training sessions.

Sometime in May 2017 the site went down. I searched for reports about why this might have happened, but couldn't find anything. Inquiries to the authors went unanswered. This happened just as I was gearing up for another training. Luckily it had been captured by Internet Archive and I directed students to use the Wayback Machine, but page loads were pretty slow. 

This project is my attempt to extract the relevant data from the archived pages. My immediate goal is to have the data available for colleagues, students, and trainees to access. My longer-term goal is to make the data available through an API, enabling others to build new resources from it (like maybe reviews, case studies, or more recent research). My stretch goal is to find a valid doi corresponding to each citation.

## How?

1. Bibliography
    1. Run `bibliography.py` to scrape the bibliography page (on Wayback) into `temp.json`.
    2. Inspect `temp.json` and make manual changes, such as joining journal name and volume, or appending missing text to citation.
    3. Detect blocks of text containing multiple citations entries.
    4. Mark separation between citations with ` || `.
    5. Run `split-blocks.py` to split text into lists of citations within each block; save into `blocks.json`.
    6. Detect and fix discrepancies between citation list and journal and href lists within each block.
    7. Run `split-citations.py` to create a separate bibliographic entry for each citation; save into `citations.json`
    8. Run `annotations.py` to scrape annotation text from Internet Archive; save into `annotations.json`
    9. Run `separate-title.py` to extract title text into new field; save into `bibliography.json`.
2. Topic pages
    1. Run `scrape-topics.py` to scrape the topic pages (on Wayback) into `topics.json`.
    2. Inspect `topics.json` and make manual changes, such as paragraph breaks and missing subtopics
3. Validation
    - Check that each link in the topic pages corresponds to an `href` in either the bibliography entries or to a topic page.

## To do

- [x] Scrape topic pages and store as JSON
- [x] Check manually
- [x] Validate links
- [ ] Add attribution

## Reflection

This was one of the more challenging scraping projects I've done in a while. I'm glad I invested time to learn Xpath syntax; I really needed it for navigating the structures in this website. Regular expressions expedited some of the parsing, too. I appreciated Python's extensive support especially the mercy that is the `re.escape` function!

The website overall had a reasonable high-level structure, divided between topics and bibliography. A lot of the challenge arose from the text itself. Not every paragraph fell into a `<p>` node, and around any inline element (`span`, `a`, formatting) was a veritable alphabet soup of various white-space characters. Unfortunately there was no clear pattern to the number or sequence of these characters, so I couldn't use them to infer paragraph breaks. My best guess is that the text was originally written in some word processor like MS Word, and then copied and pasted into a CMS interface. The process likely would have included invisible formatting codes that the CMS then translated into these excess characters.

I learned to embrace regular expressions, having resisted or written around them for the past 15 years. Another thing I discovered was that you can unpack lists much as you would tuples; not sure how I missed that for the past 20 years. And this is not new, but I noticed that I continue to favor comprehensions for list and dictionary creation. This became apparent when writing the scanning routine in `scrape-topics.py`, which really called for a more imperative pattern.


changelog.md

# Scrape bibliography page

Run bibliography scraper.

```bash
    python scrape.py > scrape.json
```

Resulting document has 170 blocks, with following fields:

```
    index:
    journals:
    -
    links:
    - text:
        href:
    text:
```

# Manual changes to journal lists and text as needed

```
    15.journals: Add name "Journal of Experimental Social Psychology, 40"
    42.journals: Change "oping" to "Coping"
    59.journals: remove empty 4th journal
    67.journals: Add name "Sex Roles, 47"
    67.text: append " Blatant stereotype threat and women’s math performance: Self-handicapping as a strategic means to cope with obtrusive negative performance expectations. Sex Roles, 47, 193–198."
    77.journals: join journal name and volume
    82.journals: join journal name and volume
    83.journals: join journal name and volume
    98.journals: add name "Journal of Applied Social Psychology, 32"
    98.text: append " Stereotype threat effects on the Raven Advanced Progressive Matrices scores of African Americans. Journal of Applied Social Psychology, 32,767–787."
    101.journals: add name "Journal of Personality and Social Psychology, 83"
    101.text: append " Journal of Personality and Social Psychology, 83, 896–918."
    107.journals: remove empty 2nd journal
    115.journals: remove empty 2nd journal
    117.journals: join journal name and volume
    120.journals: remove empty 7th journal
    124.journals: remove empty 2nd journal
    145.journals: join journal name and volume
    158.journals[1]: join journal name and volume
    158.journals[6]: join journal name and volume
    163.journals: join journal name and volume
    167.journals: remove empty 2nd journal
    167.journals: remove empty 4th journal
```

Save changes to `scrape.json`.


# Manual changes to separate text in blocks

First find out which blocks have multiple entries. Seems that ")." is a pretty consistent marker of a unique citation. Search for this in the 'text' field of each block and fitler on it.

```python
    # in python shell
    %run scrape.py
    block_counts = filter( lambda (i,b): b > 1, [ (i, block['text'].count(').')) for i,block in enumerate(blocks) ] )
```

Result is:
```
    (30, 2) - not really
    (31, 4) - only 2
    (39, 2) 
    (42, 7)
    (48, 5)
    (52, 3)
    (56, 2) - not really
    (59, 4)
    (80, 2) - not really
    (90, 2)
    (118, 3)
    (119, 2)
    (120, 13)
    (135, 2)
    (144, 2)
    (148, 2)
    (150, 2)
    (154, 2)
    (158, 8)
    (167, 2)
```

Open `scrape.json` and add the character sequence "` || `" between citations in blocks that contain multiple citations. Save changes to `scrape.json`


# Split blocks

For each string in the 'text' field with a list of strings, split across the separator ` || ` and store result as the property `citations`.

Run `python split-blocks.py > blocks.json` 

```python
# in file: split-blocks.py
    with open('scrape.json', 'r') as f:
        blocks = json.load(f)
    
    for block in blocks:
        block['citations'] = block['text'].split(" || ")
```

Also checks for and removes empty journal entries. 


# Check for correspondence

Check that every journal in a block can be found in its corresponding citation list. Check that every link text in a block can be found in its corresponding citation list.


```python
    # in python shell
    %run split-blocks.py
    journal_check = [ 
        (block['index'], [
            True in [
                (journal in citation)
                for citation in block['citations']
            ]
            for journal in block['journals']
        ] )
        for block in blocks
    ] # generates list of tuples e.g. (144, [True, True])

    link_check = [ 
        (block['index'], [
            True in [
                (link['text'] in citation)
                for citation in block['citations']
            ]
            for link in block['links']
        ] )
        for block in blocks
    ] # generates list of tuples e.g. (144, [True, True])
```

Detect any discrepancies.

```python
    # in python shell
    [ i for (i, checks) in journal_check if True not in checks ]
    # [15, 131, 168, 169]

    [ i for (i, checks) in link_check if False in checks ]
    # []
```

No issues for link correspondence! But for journal correspondence, four blocks contain `False` or no values. 

# Manual changes to resolve inconsistencies in journals

```
    15.citations: is incomplete. Append " When saying and doing diverge: The effects of stereotype threat on self-reported versus non-verbal anxiety. Journal of Experimental Social Psychology, 40, 247–255."
    131: is empty. Delete
    168: is a duplicate. Delete
    169: is empty. Delete
    198: is a duplicate. Delete
    159: is a duplicate. Delete (discovered by inspection)
```

Save changes to `blocks.json`.

# Next

At this point, we should be able to construct a list of individual citations:
```
    - citation: 
      citation_md:  # marked up with <i class='journal'></i>
      href:         # use in the future just as an identifier, or to construct separate file
      title:        # scraped from wayback annotation page
      annotation:   # scraped from wayback
      doi:          # much bigger project probably using Google Scholar API
```

# Split to individual citations

Run `python split-citations.py > citations.json` 

Also checks for and removes duplicates, using `pandas.DataFrame.groupby` to identify unique groups.

In `citations.json`, add entry for Abrams, Eller, and Bryant b/c wasn't picked up in initial scrape.

# Following links to grab annotations

Check that each `href` property in citations starts with the word "bibliography", indicating a relative link.

```python
    [ 
        (i, citation['href'][0:12]) 
        for i, citation in enumerate(citations) 
        if len(citation['href']) > 0 and citation['href'][0:12] != "bibliography" 
    ]
    #  [(86, u'https://web.')]
```


split-citations.py
import json
from pandas import DataFrame

with open('blocks.json', 'r') as f:
    blocks = json.load(f)

bibliography = []

for block in blocks:
    for citation in block['citations']:

        href = ""
        for link in block['links']:
            if link['text'] in citation:
                href = link['href'][0] 
                # [0]  is to get rid of list structure mistakenly introduced in bibliography.py

        journal = ""
        for journal_text in block['journals']:
            if journal_text in citation:
                journal = journal_text
        
        bibliography.append(
            {
                'citation' : citation,
                'journal' : journal,
                'href' : href,
            }
        )

# # add an index just for testing
# for i, each in enumerate(bibliography):
#     each['index'] = i

# eliminate duplicates
df = DataFrame(bibliography)
groups = df.groupby('citation').groups # dict of unique group : index
indices = [ index[0] for group,index in groups.items() ] 
indices.sort()
citations = [ bibliography[i] for i in indices ]

print json.dumps(citations, indent=2)


topics/scrape-topics.py
from urllib2 import urlopen
from lxml import html, etree
import json
import re

base_url = 'https://web.archive.org/web/20170426143954/http://www.reducingstereotypethreat.org:80/'

# make a dict mapping each url to the question
url = base_url
doc = html.parse(urlopen(url)).getroot()
page = doc.xpath('/html/body/table/tr/td/table/tr[5]/td/table')[0]

questions = { a.xpath('.//@href')[0] : a.xpath('./img/@alt')[0] for a in page.xpath('.//a') }

# collect hrefs, topics, and questions
url = base_url + 'definition.html'
doc = html.parse(urlopen(url)).getroot()
page = doc.xpath('/html/body/table/tr/td/table/tr[3]/td/table/tr')[0]

topics = [ 
    { 
        'href' : a.xpath('.//@href')[0],
        'topic' : a.xpath('./img/@alt')[0],
        'question' : questions[a.xpath('.//@href')[0]],
    }
    for a in page.xpath('.//a') 
]


def relativize(link):
    pattern = u'https\\:\\/\\/.*www\\.reducingstereotypethreat\\.org\\/'
    return re.sub(pattern, "", link)    

def cleanup(s):
    return ' '.join(s.split())


for topic in topics:
    url = base_url + topic['href']
    doc = html.parse(urlopen(url)).getroot()
    page = doc.xpath('/html/body/table/tr/td/table/tr[6]/td/table/tr[2]/td')[0]

    # check for nav table; if it exists, remove from parent
    try:
        nav = page.xpath('.//span/table')[0]
        nav.getparent().getparent().remove(nav.getparent())
    except IndexError:
        pass

    # get links and headings from tags in text, then clean up text

    links = [ 
        {
            'href' : relativize(a.attrib['href']), 
            'text' : cleanup(a.text_content())
        } 
        for a in page.xpath('.//a[@href]') 
        if not "#top" in a.attrib['href']
    ]
    
    subtopics = [
        {
            'name' : a.attrib['name'],
            'title' : cleanup(a.getparent().text_content())
        }
        for a in page.xpath('.//a[@name]') 
    ]

    text = cleanup(page.text_content())

    # scan for link text, replace with Markdown-style link tags

    scanned_text = []
    unscanned_text = text

    for link in links:
        [ left, right ] = unscanned_text.split(link['text'], 1)  # maxsplit = 1
        center = u"[{text}]({href})".format(**link)
        scanned_text.append(left)
        scanned_text.append(center)
        unscanned_text = right

    scanned_text.append(unscanned_text)
    text = ''.join(scanned_text)


    # scan for heading text, break into subtopic blocks

    scanned_text = []
    unscanned_text = text

    for subtopic in subtopics:
        [ left, right ] = unscanned_text.split(subtopic['title'], 1)
        scanned_text.append(left)
        unscanned_text = right

    last_text = re.sub("[Bb]ack [Tt]o [Tt]op", "", unscanned_text)
    scanned_text.append(last_text)

    for i, subtopic in enumerate(subtopics):
        subtopic['text'] = scanned_text[i + 1]

    topic['text'] =  scanned_text[0]
    topic['subtopics'] = subtopics

print json.dumps(topics, indent=2)


bibliography/annotations.py
from urllib2 import urlopen, HTTPError
from lxml import html
from time import sleep
import json

base_url = 'https://web.archive.org/web/20160609033304/http://reducingstereotypethreat.org:80/'

with open("citations.json", "r") as f:
    citations = json.load(f)

def extract_annotation(doc):
    elem = doc.xpath('/html/body/table/tr/td/table/tr[6]/td/table/tr[1]/td')[0]
    nav_links = elem.xpath('.//a')
    for nav in nav_links:
        nav_text = nav.text_content().lower()
        if ("back" in nav_text) or ("previous" in nav_text):
            nav.getparent().remove(nav)
    
    return elem.text_content()

def cleanup(s):
    return ' '.join(s.split())

for citation in citations:
    sleep(2)
    href = citation['href']

    if len(href) > 0:
        try:
            doc = html.parse(urlopen(base_url + href)).getroot()    
            citation['annotation'] = cleanup(extract_annotation(doc))
        except HTTPError:
            print "not found: " + href

print json.dumps(citations, indent=2)


scrape.py
from urllib2 import urlopen
from lxml import html, etree
import json

url = 'https://web.archive.org/web/20160609033304/http://reducingstereotypethreat.org:80/bibliography.html'
doc = html.parse(urlopen(url)).getroot()
page = doc.xpath('/html/body/table/tr/td/table/tr[6]/td/table/tr[2]/td')[0]

paragraphs = page.xpath('.//p')
# Some of these paragraphs enclose multiple <p> elements, and some of these are blank lines
# But if there is a blank line, it likely doesn't have a corresponding <a>

def cleanup(s):
    return ' '.join(s.split())

blocks = [
    {
        'index' : i,
        'text' : cleanup(p.text_content()),
        'links' : [
            {
                'href' : link.xpath('.//@href'),
                'text' : cleanup(link.text_content()),
            } for link in p.xpath('.//a') 
        ],
        'journals' : [ cleanup(i.text_content()) for i in p.xpath('.//i') ],
    } 
    for i,p in enumerate(paragraphs)
]

print json.dumps(blocks, indent=2)


    

# Bosson (2004) missing journal ("Journal of Experimental Social Psychology, 40")
# Dweck (1999) name of journal "oping" should be "Coping"
# Inzilcht, 4th journal is empty
# Keller (2000) missing journal ("Sex Roles")
# Kray (2002) journal and volume split
# Levy (1996) journal and volume split
# Leyens (2000) journal and volume split
# McKay (2002) missing journal "Journal of Applied Social Psychology, 32"
# Mendoza (2002) missing journal "Journal of Personality and Social Psychology, 83"
# Nguyen (2008) journal and volume split
# Osborne (2007) journal and volume split
# Osborne (2000) journal and volume split
# Rydell (in press) journal and volume split
# Rosenthal (2007) journal and volume split
# Ryan (2005) duplicate
# Sackett (2004) duplicate
# Sackett (2001) duplicate
# Schmader (2008) duplicate
# Stangor (1998) journal and volume split
# Walsh (1999) journal and volume split
# Wheeler (2001) journal and volume split; duplicate
# Wheeler (2001) journal and volume split; duplicate
# Yopyk (2005) journal and volume split
# Zirkel (2004) journal and volume split

